c2rust-analyze/src/main.rs

```diff
@@ -99,6 +104,250 @@ impl<T> DerefMut for MaybeUnset<T> {
+            the metadata gathered for each struct reaches a fixed point.
+        */
+        loop_count += 1;
+
+        eprintln!("---- running fixed point struct field analysis iteration #{loop_count:?} ----");
+        let old_adt_metadata = adt_metadata_table.table.clone();
+        let mut next_hypo_origin_id = 0;
+
+        // for each struct, gather lifetime information (actual and hypothetical)
+        for struct_did in &adt_metadata_table.struct_dids {
+            let adt_def = tcx.adt_def(struct_did);
+            eprintln!("gathering lifetimes and lifetime parameters for {adt_def:?}");
+            for field in adt_def.all_fields() {
+                let field_ty: Ty = tcx.type_of(field.did);
+                eprintln!("\t{adt_def:?}.{:}", field.name);
+                // iterate over dereferencable field types, e.g. &mut *mut *const &const i32
+                // and account for each in this struct's lifetime parameters
+                let mut fully_derefed_field_ty = field_ty;
+                loop {
+                    match fully_derefed_field_ty.kind() {
+                        TyKind::RawPtr(ty) => {
+                            eprintln!(
+                                "\t\tfound pointer that requires hypothetical lifetime: *{:}",
+                                if let Mutability::Mut = ty.mutbl {
+                                    "mut"
+                                } else {
+                                    "const"
+                                }
+                            );
+                            adt_metadata_table
+                                .table
+                                .entry(*struct_did)
+                                .and_modify(|adt| {
+                                    let origin_arg = OriginArg::Hypothetical(next_hypo_origin_id);
+                                    let origin_param =
+                                        OriginParam::Hypothetical(next_hypo_origin_id);
+                                    eprintln!(
+                                        "\t\t\tinserting origin {origin_param:?} into {adt_def:?}"
+                                    );
+
+                                    adt.lifetime_params.insert(origin_param);
+                                    next_hypo_origin_id += 1;
+                                    adt.field_info
+                                        .entry(field.did)
+                                        .or_default()
+                                        .lifetime
+                                        .insert(origin_arg);
+                                });
+                            fully_derefed_field_ty = ty.ty;
+                        }
+                        TyKind::Ref(reg, ty, _mutability) => {
+                            eprintln!("\t\tfound reference field lifetime: {reg:}");
+                            let origin_arg = OriginArg::Actual(*reg);
```

Comment:
Can you add assertions here and in the ADT case to make sure that `ReEarlyBound` and `ReStatic` are the only things that can be placed in the `OriginArg::Actual` variant?  I'm pretty sure those are the only two `RegionKind`s that can appear inside an `AdtDef`, but it would be good to have a check in case I'm wrong or in case rustc's behavior ever changes.

Reply:
added in d8e813cb, let me know if I've missed any cases to check

Reply:
Looks good - those seem to be the only places where `OriginArg::Actual` is constructed.